tony-core/src/main/java/com/linkedin/tony/TonyClient.java

```diff
@@ -390,22 +390,57 @@ public boolean init(String[] args) throws ParseException, IOException {
       shellEnv.putAll(Utils.parseKeyValue(envs));
     }
 
+    if (!parseDockerConfigs()) {
+      return false;
+    }
+
+    if (cliParser.hasOption("container_env")) {
+      String[] containerEnvs = cliParser.getOptionValues("container_env");
+      containerEnv.putAll(Utils.parseKeyValue(containerEnvs));
+    }
+
+    return true;
+  }
+
+  /**
+   * Parses Docker related configs and sets the appropriate container environment variables if Docker is available.
+   * Uses reflection to support older versions of Hadoop.
+   */
+  private boolean parseDockerConfigs() {
```

Comment:
This makes me think twice on merging this into master branch. There is gonna be more and more docker related featured been developed in Hadoop 3.x, this is gonna make future development miserable..

Reply:
I think we can make it maintainable by keeping Docker-related methods in one place -- perhaps creating a DockerUtils class.

Also, we already support Docker -- we just rely on YARN being configured to use the Docker container runtime and users specifying a Docker image via config.

What other Docker work is needed?

Reply:
Another alternative if we want to keep building against a newer Hadoop version is to use the `hadoop27` branch to publish a separate 2.7-compatible internal TonY build that can run on our Hadoop 2.7 clusters. We could use version numbers to differentiate the 2, e.g.: 0.1.x for TonY Hadoop 2.7 releases and 0.2.x+ for TonY Hadoop 2.9 releases.

What do you think we should do?

Reply:
@erwa I think an Executor config or class to abstract out things would be useful. On a second thought, if we want to run TonY in 2.7 environment, it is much better to have single version to support both, otherwise, users would have to build two MPs one for adhoc cluster and one for prod cluster.

Reply:
Yeah, we're in alignment that having a single version that supports both 2.7 and 2.9 makes more sense.

I think we can do an "Executor config" refactoring in a separate PR.